romanisim/cr.py

```diff
@@ -0,0 +1,326 @@
+    jj = np.concatenate((jj, [np.round(j1).astype(int)]))
+
+    # if no crossings, then it's just the total Euclidean distance
+    if len(crossings) == 0:
+        lengths = np.linalg.norm([di, dj])
+    # otherwise, compute starting, crossing, and ending distances
+    else:
+        first_length = np.linalg.norm(crossings[0] - np.array(trail_start),
+                                      keepdims=1)
+        middle_lengths = np.linalg.norm(np.diff(crossings, axis=0), axis=1)
+        last_length = np.linalg.norm(np.array(trail_end) - crossings[-1],
+                                     keepdims=1)
+        lengths = np.concatenate([first_length, middle_lengths, last_length])
+
+        # remove 0-length trails
+        positive_lengths = lengths > 0
+        lengths = lengths[positive_lengths]
+        ii = ii[positive_lengths]
+        jj = jj[positive_lengths]
+
+    return ii, jj, lengths
+
+
+def simulate_crs(image, time, flux=8, area=0.168, gain=3.8, pixel_size=10,
+                 pixel_depth=5, rng=None, seed=47):
+    """Adds CRs to an existing image.
+
+    Parameters
+    ----------
+    image : 2-d array of floats
+        The detector image with values in units of counts.
+    time : float
+        The exposure time, units of s.
+    flux : float
+        Cosmic ray flux, units of cm^-2 s^-1. Default value of 8
+        is equal to the value assumed by the JWST ETC.
+    area : float
+        The area of the WFI detector, units of cm^2.
+    gain : float
+        The gain to convert from eV to counts, units of eV / counts.
+    pixel_size : float
+        The size of an individual pixel in the detector, units of micron.
+    pixel_depth : float
+        The depth of an individual pixel in the detector, units of micron.
+    rng : np.random.Generator
+        Random number generator to use
+    seed : int
+        seed to use for random number generator
+
+    Returns
+    -------
+    image : 2-d array of floats
+        The detector image, in units of counts, updated to include
+        all of the generated cosmic ray hits.
+    """
+
+    if rng is None:
+        rng = np.random.default_rng(seed)
+
+    N_i, N_j = image.shape
+    N_samples = rng.poisson(flux * area * time)
+    cr_i0, cr_j0, cr_angle, cr_length, cr_dEdx = sample_cr_params(
+        N_samples, N_i=N_i, N_j=N_j, rng=rng)
+
+    cr_length = cr_length / pixel_size
+    cr_i1 = (cr_i0 + cr_length * np.cos(cr_angle)).clip(1, N_i - 1)
+    cr_j1 = (cr_j0 + cr_length * np.sin(cr_angle)).clip(1, N_j - 1)
+
+    # go from eV/micron -> counts/pixel
+    cr_counts_per_pix = cr_dEdx * pixel_size / gain
+
+    for i0, j0, i1, j1, counts_per_pix in zip(
+            cr_i0, cr_j0, cr_i1, cr_j1, cr_counts_per_pix):
+        ii, jj, length_2d = traverse([i0, j0], [i1, j1])
+        length_3d = ((pixel_depth / pixel_size) ** 2 + length_2d ** 2) ** 0.5
+        image[ii, jj] += rng.poisson(counts_per_pix * length_3d)
+
+    return image
+
+
+if __name__ == "__main__":
```

Comment:
Not sure if having `__main__` here is intended, but if it is, it would be good to have the ability to change the hardcoded values below at runtime

Reply:
John had these in to have some simple tests for the CR module.  I can remove those after we build out the CR tests; let's not worry about that here.

Reply:
Got it. Thanks!